engine/pkg/election/elector.go

```diff
@@ -0,0 +1,357 @@
+
+	// Add self to the record if not exists.
+	if m, ok := record.FindMember(e.config.ID); !ok {
+		record.Members = append(record.Members, &Member{
+			ID:            e.config.ID,
+			Name:          e.config.Name,
+			Address:       e.config.Address,
+			LeaseDuration: e.config.LeaseDuration,
+			RenewTime:     time.Now(),
+		})
+	} else {
+		m.RenewTime = time.Now()
+	}
+
+	if time.Now().Before(e.resignUntil) {
+		if record.LeaderID == e.config.ID {
+			record.LeaderID = ""
+			log.Info("resign the leadership")
+		}
+	} else if record.LeaderID == "" {
+		// Elect a new leader if no leader exists.
+		record.LeaderID = e.config.ID
+		log.Info(
+			"try to elect self as leader",
+			zap.String("id", e.config.ID),
+			zap.String("name", e.config.Name),
+			zap.String("address", e.config.Address),
+		)
+	}
+
+	if err := s.Update(ctx, record); err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record)
+	return nil
+}
+
+func (e *Elector) release(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, defaultReleaseTimeout)
+	defer cancel()
+
+	s := e.config.Storage
+
+	record, err := s.Get(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record.Clone())
+
+	for i, m := range record.Members {
+		if m.ID == e.config.ID {
+			record.Members = append(record.Members[:i], record.Members[i+1:]...)
+			break
+		}
+	}
+	if record.LeaderID == e.config.ID {
+		record.LeaderID = ""
+	}
+
+	if err := s.Update(ctx, record); err != nil {
+		return errors.Trace(err)
+	}
+	e.setObservedRecord(record)
+	return nil
+}
+
+func (e *Elector) setObservedRecord(record *Record) {
+	e.observeLock.Lock()
+	defer e.observeLock.Unlock()
+
+	// Remove members that are not in the new record.
+	for id := range e.observedRenews {
+		if _, ok := record.FindMember(id); !ok {
+			delete(e.observedRenews, id)
+		}
+	}
+
+	// Update observedRenews for members in the new record.
+	for _, m := range record.Members {
+		oldMember, ok := e.observedRecord.FindMember(m.ID)
+		// If the member is not in the old record, or the RenewTime is
+		// changed, update the local observedRenews.
+		if !ok || !oldMember.RenewTime.Equal(m.RenewTime) {
+			e.observedRenews[m.ID] = time.Now()
+		}
+	}
+
+	// New leader is elected.
+	if record.LeaderID != "" && record.LeaderID != e.observedRecord.LeaderID {
+		leader, ok := record.FindMember(record.LeaderID)
+		if ok {
+			log.Info(
+				"new leader elected",
+				zap.String("leader-id", leader.ID),
+				zap.String("leader-name", leader.Name),
+				zap.String("leader-address", leader.Address),
+			)
+		}
+	}
+
+	e.observedRecord = *record
+}
+
+func (e *Elector) isLeaseExpired(memberID string) bool {
+	e.observeLock.RLock()
+	defer e.observeLock.RUnlock()
+
+	return e.isLeaseExpiredLocked(memberID)
+}
+
+func (e *Elector) isLeaseExpiredLocked(memberID string) bool {
+	member, ok := e.observedRecord.FindMember(memberID)
+	if !ok {
+		return true
+	}
+	renewTime := e.observedRenews[memberID]
```

Comment:
Can we ensure `observedRenews` always has the same members with `observedRecord`

Reply:
Yes, it's guaranteed in `setObservedRecord`.